File: apps/api/src/allrounder_api/contracts.py
from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, HttpUrl


def alias_config() -> ConfigDict:
    return ConfigDict(
        alias_generator=lambda value: _to_camel(value),
        populate_by_name=True,
        extra="forbid",
    )


def _to_camel(value: str) -> str:
    first, *rest = value.split("_")
    return first + "".join(word.title() for word in rest)


class Contract(BaseModel):
    model_config = alias_config()


class Domain(StrEnum):
    CODE = "code"
    FINANCE = "finance"
    MARKETING = "marketing"
    SUPPORT = "support"
    UNKNOWN = "unknown"


class Gate(StrEnum):
    AUTO = "auto"
    APPROVAL = "approval"
    REFUSE = "refuse"


class Attachment(Contract):
    id: str = Field(min_length=1)
    name: str = Field(min_length=1)
    mime: str = Field(min_length=1)
    bytes: int = Field(ge=0)


class Ticket(Contract):
    key: str = Field(pattern=r"^[A-Z][A-Z0-9_]*-\d+$")
    project: str = Field(min_length=1)
    issue_type: str = Field(min_length=1)
    labels: list[str]
    priority: str = Field(pattern=r"^(Highest|High|Medium|Low|Lowest)$")
    summary: str = Field(min_length=1, max_length=500)
    description: str
    reporter: str = Field(min_length=1)
    attachments: list[Attachment]
    event_id: str = Field(min_length=1)
    received_at: datetime


class TriageVerdict(Contract):
    domain: Domain
    confidence: float = Field(ge=0, le=1)
    urgency: int = Field(ge=1, le=5)
    needs_human: bool
    rationale: str = Field(min_length=1)


class RiskScore(Contract):
    action: str = Field(min_length=1)
    blast_radius: str = Field(pattern=r"^(low|med|high)$")
    reversibility: str = Field(pattern=r"^(reversible|compensable|irreversible)$")
    score: int = Field(ge=0, le=100)
    gate: Gate
    reasons: list[str]


class ActionEvidence(Contract):
    tool: str
    args_redacted: dict[str, object]
    status: str
    artifact_urls: list[HttpUrl]


class Citation(Contract):
    source_id: str
    span: str


class ModelUse(Contract):
    step: str
    model: str
    tokens: int = Field(ge=0)


class EvidencePack(Contract):
    actions: list[ActionEvidence]
    citations: list[Citation]
    confidence: float = Field(ge=0, le=1)
    cost_usd_micro: int = Field(ge=0)
    transcript_ref: str = Field(min_length=1)
    model_trail: list[ModelUse]


class RoutedTicket(Contract):
    ticket: Ticket
    verdict: TriageVerdict
    risk_scores: list[RiskScore]
    gate: Gate
    workflow: str

File: src/mastra/shared/contracts.ts
import { z } from "zod";

export const DomainSchema = z.enum(["code", "finance", "marketing", "support", "security", "unknown"]);
export const GateSchema = z.enum(["auto", "approval", "refuse"]);

export const AttachmentSchema = z
  .object({
    id: z.string().min(1),
    name: z.string().min(1),
    mime: z.string().min(1),
    bytes: z.number().int().nonnegative(),
  })
  .strict();

export const TicketSchema = z
  .object({
    key: z.string().regex(/^[A-Z][A-Z0-9_]*-\d+$/),
    project: z.string().min(1),
    issueType: z.string().min(1),
    labels: z.array(z.string()),
    priority: z.enum(["Highest", "High", "Medium", "Low", "Lowest"]),
    summary: z.string().min(1).max(500),
    description: z.string(),
    reporter: z.string().min(1),
    attachments: z.array(AttachmentSchema),
    eventId: z.string().min(1),
    receivedAt: z.string().datetime({ offset: true }),
  })
  .strict();

export const TriageVerdictSchema = z
  .object({
    domain: DomainSchema,
    confidence: z.number().min(0).max(1),
    urgency: z.number().int().min(1).max(5),
    needsHuman: z.boolean(),
    rationale: z.string().min(1),
  })
  .strict();

export const RiskScoreSchema = z
  .object({
    action: z.string().min(1),
    blastRadius: z.enum(["low", "med", "high"]),
    reversibility: z.enum(["reversible", "compensable", "irreversible"]),
    score: z.number().int().min(0).max(100),
    gate: GateSchema,
    reasons: z.array(z.string()),
  })
  .strict();

export const EvidencePackSchema = z
  .object({
    actions: z.array(
      z
        .object({
          tool: z.string(),
          argsRedacted: z.record(z.unknown()),
          status: z.string(),
          artifactUrls: z.array(z.string().url()),
        })
        .strict(),
    ),
    citations: z.array(
      z.object({ sourceId: z.string(), span: z.string() }).strict(),
    ),
    confidence: z.number().min(0).max(1),
    costUsdMicro: z.number().int().nonnegative(),
    transcriptRef: z.string().min(1),
    modelTrail: z.array(
      z
        .object({
          step: z.string(),
          model: z.string(),
          tokens: z.number().int().nonnegative(),
        })
        .strict(),
    ),
  })
  .strict();

export type Domain = z.infer<typeof DomainSchema>;
export type Gate = z.infer<typeof GateSchema>;
export type Ticket = z.infer<typeof TicketSchema>;
export type TriageVerdict = z.infer<typeof TriageVerdictSchema>;
export type RiskScore = z.infer<typeof RiskScoreSchema>;
export type EvidencePack = z.infer<typeof EvidencePackSchema>;

